packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -892,6 +892,18 @@ function readFromUnsubcribedMutableSource<Source, Snapshot>(
   const getVersion = source._getVersion;
   const version = getVersion(source._source);
 
+  let mutableSourceSideEffectDetected = false;
+  if (__DEV__) {
+    // Detect side effects that update a mutable source during render.
+    // See https://github.com/facebook/react/issues/19948
+    if (source._currentlyRenderingFiber !== currentlyRenderingFiber) {
+      source._currentlyRenderingFiber = currentlyRenderingFiber;
+      source._initialVersionAsOfFirstRender = version;
```

Comment:
This will cause us to over-retain Fibers a little DEV mode. I assume that's not a problem but if it was I could add explicit cleanup somewhere.